File: shared/src/commonMain/kotlin/dev/chara/tasks/firebase/Firebase.kt
package dev.chara.tasks.firebase

object Firebase {
    private lateinit var instance: FirebaseWrapper

    fun link(platformWrapper: FirebaseWrapper) {
        instance = platformWrapper
    }

    suspend fun getMessagingToken(): String? = instance.getMessagingToken()
    fun setUserId(userId: String) = instance.setUserId(userId)
    fun log(msg: String) = instance.log(msg)
    fun recordException(t: Throwable) = instance.recordException(t)
}

interface FirebaseWrapper {
    suspend fun getMessagingToken(): String?
    fun setUserId(userId: String)
    fun log(msg: String)
    fun recordException(t: Throwable)
}

File: ios/Tasks/Firebase/Firebase.swift
//
//  Firebase.swift
//  Tasks
//
//  Created by Luke Myers on 7/20/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import FirebaseCore
import FirebaseCrashlytics
import FirebaseMessaging
import Foundation
import TasksShared

class AppleFirebaseWrapper: FirebaseWrapper {
    func getMessagingToken() async throws -> String? {
        Messaging.messaging().fcmToken
    }

    func log(msg: String) {
        Crashlytics.crashlytics().log(msg)
    }

    func recordException(t: KotlinThrowable) {
        CrashlyticsKt.recordException(throwable: t)
    }

    func setUserId(userId: String) {
        Crashlytics.crashlytics().setUserID(userId)
    }
}
